src/file_loader.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2013-2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib::object::IsA;
+use glib::translate::*;
+use gtk_source_sys;
+use std::boxed::Box;
+use std::ptr;
+use FileLoader;
+
+pub trait FileLoaderExtManual {
+    fn load_async<
+        P: IsA<gio::Cancellable>,
+        Q: 'static + Fn(i64, i64) + Send,
+        R: 'static + Fn(Result<(), glib::Error>) + Send,
+    >(
+        &self,
+        io_priority: glib::Priority,
+        cancellable: Option<&P>,
+        progress_callback: Q,
+        ready_callback: R,
+    );
+}
+
+impl<O: IsA<FileLoader>> FileLoaderExtManual for O {
+    fn load_async<
+        P: IsA<gio::Cancellable>,
+        Q: 'static + Fn(i64, i64) + Send,
+        R: 'static + Fn(Result<(), glib::Error>) + Send,
+    >(
+        &self,
+        io_priority: glib::Priority,
+        cancellable: Option<&P>,
+        progress_callback: Q,
+        ready_callback: R,
+    ) {
+        unsafe extern "C" fn load_async_progress_trampoline<
+            Q: 'static + Fn(i64, i64) + Send,
+            R: 'static + Fn(Result<(), glib::Error>) + Send,
+        >(
+            current_num_bytes: i64,
+            total_num_bytes: i64,
+            user_data: glib_sys::gpointer,
+        ) {
+            let callbacks = &*(user_data as *mut (Q, R));
+            callbacks.0(current_num_bytes, total_num_bytes);
+        }
+
+        unsafe extern "C" fn load_async_ready_trampoline<
+            Q: 'static + Fn(i64, i64) + Send,
+            R: 'static + Fn(Result<(), glib::Error>) + Send,
+        >(
+            source_object: *mut gobject_sys::GObject,
+            res: *mut gio_sys::GAsyncResult,
+            user_data: glib_sys::gpointer,
+        ) {
+            let mut error = ptr::null_mut();
+            let ret = gtk_source_sys::gtk_source_file_loader_load_finish(
+                source_object as *mut _,
+                res,
+                &mut error,
+            );
+            let result = if !error.is_null() {
+                Err(from_glib_full(error))
+            } else if ret == 0 {
+                Err(glib::Error::new(
+                    glib::FileError::Failed,
+                    "Unexpected failure",
+                ))
+            } else {
+                Ok(())
+            };
+            let callbacks = &*(user_data as *mut (Q, R));
+            callbacks.1(result);
+        }
+
+        unsafe extern "C" fn load_async_notify_trampoline<
+            Q: 'static + Fn(i64, i64) + Send,
+            R: 'static + Fn(Result<(), glib::Error>) + Send,
+        >(
+            user_data: glib_sys::gpointer,
+        ) {
+            Box::from_raw(user_data as *mut _);
```

Comment:
I'm not sure what exactly this does but this at least leaks memory, if not even causing memory corruption. You always have to specify the concrete type of the box here

Reply:
Hmm... that's an interesting one. Why doesn't the compiler flag it, if it's known to be bad? Perhaps because it's in an unsafe block?

Reply:
Yes, a) because it's unsafe code so it's your job to ensure that raw pointers are used correctly and b) because of the `_` cast the target type is whatever the compiler wants here